File: 2025/07 July/28 Make Matrix Beautiful/main.py
# Link: https://www.geeksforgeeks.org/problems/make-matrix-beautiful-1587115620/1


class Solution:
    def balanceSums(self, mat):
        n = len(mat)
        res = 0
        maxSum = 0

        for i in range(n):
            row_total = sum(mat[i])
            maxSum = max(row_total, maxSum)

        for j in range(n):
            col_total = sum(mat[i][j] for i in range(n))
            maxSum = max(col_total, maxSum)

        for i in range(n):
            row_total = sum(mat[i])
            res += maxSum - row_total

        return res


if __name__ == "__main__":
    sol = Solution()
    mat = [[1, 2], [3, 4]]
    result = sol.balanceSums(mat)
    print(f"Minimum operations to make the matrix beautiful: {result}")

    mat = [[1, 2, 3], [4, 2, 3], [3, 2, 1]]
    result = sol.balanceSums(mat)
    print(f"Minimum operations to make the matrix beautiful: {result}")

File: 2025/07 July/28 Make Matrix Beautiful/main.cpp
// Link: https://www.geeksforgeeks.org/problems/make-matrix-beautiful-1587115620/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    int balanceSums(vector<vector<int>> &mat)
    {
        int n = mat.size();
        int res = 0;
        int maxSum = 0;

        for (int i = 0; i < n; i++)
        {
            int sum = 0;
            for (int j = 0; j < n; j++)
                sum += mat[i][j];
            maxSum = max(sum, maxSum);
        }

        for (int j = 0; j < n; j++)
        {
            int sum = 0;
            for (int i = 0; i < n; i++)
                sum += mat[i][j];
            maxSum = max(sum, maxSum);
        }

        for (int i = 0; i < n; i++)
        {
            int sum = 0;
            for (int j = 0; j < n; j++)
            {
                sum += mat[i][j];
            }
            res += (maxSum - sum);
        }
        return res;
    }
};

int main()
{
    Solution sol;
    vector<vector<int>> mat = {
        {1, 2},
        {3, 4}};
    int result = sol.balanceSums(mat);
    cout << "Minimum operations to make the matrix beautiful: " << result << endl;

    mat = {
        {1, 2, 3},
        {4, 2, 3},
        {3, 2, 1}};
    result = sol.balanceSums(mat);
    cout << "Minimum operations to make the matrix beautiful: " << result << endl;
    return 0;
}